File: DSA/Linkedlist/Code/DetectCycleInDoublyLinkedList.java
// Detect Cycle in Doubly Linked List in Java

class Node {
    int data;
    Node next;
    Node prev;
    
    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

public class CycleDetectionDLL {
    Node head;
    
    void addNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
        newNode.prev = temp;
    }
    
    boolean detectCycleForward() {
        if (head == null) return false;
        
        Node slow = head;
        Node fast = head;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            
            if (slow == fast)
                return true;
        }
        
        return false;
    }
    
    boolean detectCycleBackward() {
        if (head == null) return false;
        
        Node temp = head;
        while (temp.next != null)
            temp = temp.next;
        
        Node slow = head;
        
        while (temp != null && slow != null) {
            if (temp == slow)
                return true;
            temp = temp.prev;
            if (temp == null) break;
            if (temp == slow)
                return true;
            slow = slow.next;
        }
        
        return false;
    }
    
    public static void main(String[] args) {
        CycleDetectionDLL list = new CycleDetectionDLL();
        
        System.out.println("=== Detect Cycle in Doubly Linked List ===\n");
        
        // Test 1: No cycle
        list.addNode(1);
        list.addNode(2);
        list.addNode(3);
        
        System.out.print("List 1 (No cycle): ");
        System.out.println(list.detectCycleForward() ? "Cycle detected" : "No cycle");
        
        // Test 2: With cycle (manual)
        Node head = new Node(1);
        Node n2 = new Node(2);
        Node n3 = new Node(3);
        Node n4 = new Node(4);
        
        head.next = n2;
        n2.prev = head;
        n2.next = n3;
        n3.prev = n2;
        n3.next = n4;
        n4.prev = n3;
        n4.next = n2;
        n2.prev = n4;
        
        System.out.print("List 2 (With cycle): ");
        
        Node slow = head;
        Node fast = head;
        boolean hasCycle = false;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            
            if (slow == fast) {
                hasCycle = true;
                break;
            }
        }
        
        System.out.println(hasCycle ? "Cycle detected" : "No cycle");
        
        System.out.println("\nComplexity Analysis:");
        System.out.println("Forward pointer detection: O(n) time, O(1) space");
        System.out.println("Backward pointer detection: O(n) time, O(1) space");
    }
}

File: DSA/Linkedlist/Code/DetectCycleInDoublyLinkedList.py
# Detect Cycle in Doubly Linked List in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class CycleDetectionDLL:
    def __init__(self):
        self.head = None
    
    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
        new_node.prev = temp
    
    def detect_cycle_forward(self):
        if self.head is None:
            return False
        
        slow = self.head
        fast = self.head
        
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                return True
        
        return False
    
    def detect_cycle_backward(self):
        if self.head is None:
            return False
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        slow = self.head
        
        while temp is not None and slow is not None:
            if temp == slow:
                return True
            temp = temp.prev
            if temp is None:
                break
            if temp == slow:
                return True
            slow = slow.next
        
        return False

# Test
print("=== Detect Cycle in Doubly Linked List ===\n")

# Test 1: No cycle
list1 = CycleDetectionDLL()
list1.add_node(1)
list1.add_node(2)
list1.add_node(3)

print("List 1 (No cycle):", "Cycle detected" if list1.detect_cycle_forward() else "No cycle")

# Test 2: With cycle
list2 = CycleDetectionDLL()
n1 = Node(1)
n2 = Node(2)
n3 = Node(3)
n4 = Node(4)

n1.next = n2
n2.prev = n1
n2.next = n3
n3.prev = n2
n3.next = n4
n4.prev = n3
n4.next = n2
n2.prev = n4

list2.head = n1

print("List 2 (With cycle):", "Cycle detected" if list2.detect_cycle_forward() else "No cycle")

print("\nComplexity Analysis:")
print("Forward pointer detection: O(n) time, O(1) space")
print("Backward pointer detection: O(n) time, O(1) space")
